proposals/0010-vk-buffer-ref.md

```diff
@@ -0,0 +1,270 @@
+
+For example, here is a shader using vk::RawBufferLoad(). Note the physical offset 16 hard-coded into the shader:
+
+```c++
+// struct GlobalsTest_t
+// {
+//       float4 g_vSomeConstantA;
+//       float4 g_vTestFloat4;
+//       float4 g_vSomeConstantB;
+// };
+
+struct TestPushConstant_t
+{
+      uint64_t m_nBufferDeviceAddress;  // GlobalsTest_t
+};
+
+[[vk::push_constant]] TestPushConstant_t g_PushConstants;
+
+float4 MainPs(void) : SV_Target0
+{
+    float4 vTest = vk::RawBufferLoad<float4>(g_PushConstants.m_nBufferDeviceAddress + 16);
+      
+      return vTest;
+}
+```
+
+The SPIR-V for this shader can be seen in Appendix A. Note the lack of logical context for the accessed buffer i.e. no declaration for the underlying structure GlobalsTest_t as is generated for other buffers.
+
+There is another way to use RawBufferLoad which does allow logical selection of the buffer fields, but it inefficiently loads the entire buffer to do it. See https://github.com/microsoft/DirectXShaderCompiler/issues/4986.
+
+The goal of this proposal is to have a solution that meets the following requirements:
+
+* Removes the need for having to manually or automatically generate offsets to load structured data with BufferDeviceAddress.
+* Get equivalent tooling functionality as is provided by the buffer reference feature in GLSL.  Namely, tools like RenderDoc are able to introspect the type information such that its buffer inspection and shader debugger are able to properly understand and represent the type of the data.
+* Make it possible through SPIR-V reflection to determine which members of a struct accessed by BufferDeviceAddress are statically referenced and at what offset.  This is already possible for other data like cbuffers in order for shader tooling to be able to identify which elements are used and where to put them.
+
+## Proposed solution
+
+Our solution is to add a new builtin type in the vk namespace that is a pointer to a buffer of a given type, `vk::buffer_ptr<T,A>`. The template argument `T` must be a struct. `A` must be an integer and is the alignment in bytes of the pointer. If `A` is not specified, the alignment is assumed to be 16 bytes.
+
+This new type will have the following operations
+
+* Copy assignment and copy construction - These copy the value of the pointer from one variable to another.
+* Accessing members of T - This is done using the `.` operator.
+* Conversion to uint64_t - Return the integer representation of the address it points to. This can be used for null pointer checks.
+
+Note the operations that are not allowed:
+
+* There is no default construction. Every vk::buffer_ptr<T> is either contained in a global resource (like a cbuffer, ubo, or ssbo), or it must be constructed using the copy constructor. 
```

Comment:
I think this restriction actually means these can't be placed in Unions by C++ rules:

>If a union contains a non-static data member of reference type the program is ill-formed. At most one non-static data member of a union may have a brace-or-equal-initializer. [Note: If any non-static data member of a union has a non-trivial default constructor (12.1), copy constructor (12.8), move constructor (12.8), copy assignment operator (12.8), move assignment operator (12.8), or destructor (12.4), the corresponding member function of the union must be user-provided or it will be implicitly deleted (8.4.3) for the union. — end note ]

Since HLSL doesn't yet support constructors and destructors, this restriction would mean that `vk::BufferPointer` cannot be placed inside a union.